src/wasm_udf/wasm.rs

```diff
@@ -11,22 +15,518 @@ use datafusion::{
+        CreateFunctionDataType::BIGINT => Ok(DataType::Int64),
+        CreateFunctionDataType::CHAR => Ok(DataType::Utf8),
+        CreateFunctionDataType::VARCHAR => Ok(DataType::Utf8),
+        CreateFunctionDataType::TEXT => Ok(DataType::Utf8),
+        CreateFunctionDataType::DECIMAL { precision, scale } => {
+            Ok(DataType::Decimal128(*precision, *scale))
+        }
+        CreateFunctionDataType::FLOAT => Ok(DataType::Float32),
+        CreateFunctionDataType::REAL => Ok(DataType::Float32),
+        CreateFunctionDataType::DOUBLE => Ok(DataType::Float64),
+        CreateFunctionDataType::BOOLEAN => Ok(DataType::Boolean),
+        CreateFunctionDataType::DATE => Ok(DataType::Date32),
+        CreateFunctionDataType::TIMESTAMP => {
+            Ok(DataType::Timestamp(TimeUnit::Nanosecond, None))
+        }
+    }
+}
+
+fn get_wasm_module_exported_fn<Params, Results>(
+    instance: &Instance,
+    store: &mut Store<WasiCtx>,
+    export_name: &str,
+) -> anyhow::Result<TypedFunc<Params, Results>>
+where
+    Params: wasmtime::WasmParams,
+    Results: wasmtime::WasmResults,
+{
+    instance
+        .get_typed_func::<Params, Results, _>(store, export_name)
+        .map_err(|err| {
+            anyhow::anyhow!(format!(
+                "Error getting export `{:?}`: {:?}",
+                export_name, err
+            ))
+        })
+}
+
+struct WasmMessagePackUDFInstance {
+    store: Store<WasiCtx>,
+    alloc: TypedFunc<i32, i32>,
+    dealloc: TypedFunc<(i32, i32), ()>,
+    udf: TypedFunc<i32, i32>,
+    memory: Memory,
+}
+
+impl WasmMessagePackUDFInstance {
+    pub fn new(module_bytes: &[u8], function_name: &str) -> anyhow::Result<Self> {
+        let engine = Engine::default();
+        let mut linker = wasmtime::Linker::new(&engine);
+        // Create a WASI context and put it in a Store; all instances in the store
+        // share this context. `WasiCtxBuilder` provides a number of ways to
+        // configure what the target program will have access to.
+        let wasi = WasiCtxBuilder::new().inherit_stderr().build();
+        let mut store = Store::new(&engine, wasi);
+        // Add both wasi_unstable and wasi_snapshot_preview1 WASI modules
+        wasmtime_wasi::add_to_linker(&mut linker, |s| s)?;
+        // Instantiate WASM module.
+        // TODO: handle module loading errors
+        let module = Module::from_binary(&engine, module_bytes)?;
+        // TODO: handle linker errors
+        let instance = linker.instantiate(&mut store, &module)?;
+
+        let alloc = get_wasm_module_exported_fn(&instance, &mut store, "alloc")?;
+        let dealloc = get_wasm_module_exported_fn(&instance, &mut store, "dealloc")?;
+        let udf = get_wasm_module_exported_fn(&instance, &mut store, function_name)?;
+        let memory = match instance.get_memory(&mut store, "memory") {
+            Some(mem) => Ok(mem),
+            None => Err(anyhow::anyhow!("could not find module's exportd memory")),
+        }?;
```

Comment:
Some more code golf (though I think this can be a DataFusionError instead):

```suggestion
        let memory = instance.get_memory(&mut store, "memory").ok_or(||
            anyhow::anyhow!("could not find module's exported memory")
        )?;
```